Modifications to support non-Linux OSes (tested on FreeBSD 14)

21M4TW opened the pull request:
-Modifying CMakeLists.txt to support non-Linux OSes that do not include epoll and argp. Using epoll-shim and the standalone argp libraries. Tested on FreeBSD 14
-Including <netiniet/in.h> in src/datum_protocol.c. This is required for compilation on FreeBSD 14

luke-jr commented on CMakeLists.txt:
```diff
@@ -41,6 +41,15 @@ pkg_check_modules(MICROHTTPD REQUIRED libmicrohttpd)
 pkg_check_modules(SODIUM REQUIRED libsodium)
 find_package(Threads REQUIRED)
 
+if(NOT LINUX)
```
This isn't a good way to do build conditionals. Need to check if epoll/argp work without extra libs, and check for extra libs if needed.

Noteworthy, some Linux distros do not have argp.

luke-jr commented:
Do you plan to finish this? (Note argp-standalone is now supported in master)

adozenlines commented:
I have added support for macOS to my fork; please feel free to modify it as needed. I have also added support for refreshing the status page. If it's something we can move into the main project, let me know.

TIA.

21M4TW commented:
> Do you plan to finish this? (Note argp-standalone is now supported in master)

Sorry yes I will look at it!

luke-jr commented:
Sounds like a CMake bug? https://cmake.org/cmake/help/latest/module/CheckLibraryExists.html says the empty location should search default paths, which CMake devs say should include /usr/local on FreeBSD: https://gitlab.kitware.com/cmake/cmake/-/issues/21117

21M4TW commented on CMakeLists.txt:
```diff
@@ -83,6 +83,17 @@ if(NOT HAVE_ARGP_PARSE)
+if(NOT HAVE_EPOLL_WAIT)
+	check_library_exists(epoll-shim epoll_wait "" EPOLL_SHIM)
+	if(EPOLL_SHIM)
+		pkg_check_modules(EPOLL_SHIM REQUIRED epoll-shim)
```
Ok, this is done.

21M4TW commented:
> Sounds like a CMake bug? https://cmake.org/cmake/help/latest/module/CheckLibraryExists.html says the empty location should search default paths, which CMake devs say should include /usr/local on FreeBSD: https://gitlab.kitware.com/cmake/cmake/-/issues/21117

Yes I guess it is a bug on the CMake side, and can probably be ignored for here

21M4TW commented:
Should be good now.

21M4TW commented:
> Do you want to add instructions to README and/or a CI task?

I modified the README and started working on the CI task (in progress). I also modified CMakeLists.txt to add a workaround for missing /usr/local/lib FreeBSD system directory in cmake, since it is also affecting the CI task. I will let you know once the CI task works...

21M4TW commented:
> Do you want to add instructions to README and/or a CI task?

Ok the CI task should be working fine now

luke-jr commented on CMakeLists.txt:
```diff
@@ -77,12 +77,24 @@ endif()
 check_function_exists(argp_parse HAVE_ARGP_PARSE)
 if(NOT HAVE_ARGP_PARSE)
-	check_library_exists(argp argp_parse "" ARGP)
+	if(CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
+		check_library_exists(argp argp_parse "/usr/local/lib" ARGP)
```
For example, if argp is moved to main, it would then be under /usr,  not /usr/local. Or if someone forks FreeBSD to do that (or perhaps something more complicated).

21M4TW replied:
All non-system libraries for FreeBSD are in /usr/local . It would only be in /usr/lib if argp was integrated into FreeBSD itself and the argp package disappeared, in which case the Datum README and the CI would need to be modified along with CMakeLists.txt? I just pushed a change that now look in /usr/lib in addition to /usr/local/lib if this ever becomes the case

21M4TW commented:
Sure this can also work, thanks. I committed it to the PR.

luke-jr commented:
Rebased this to the 0.2.x branch and cleaned up the commit history. Can you confirm the new branch works fine on FreeBSD?

21M4TW commented:
> Tested on Free-BSD - DATUM is unable to connect to OCEAN's prime servers.

Ok. I have only been solo mining with my setup, but I guess this is a good excuse to try this as well...

21M4TW commented:
> Tested on Free-BSD - DATUM is unable to connect to OCEAN's prime servers.

Do you get a seg fault and the following error?
2025-07-07 15:09:33.936 [                       datum_protocol_client] ERROR: connect(...) error: Permission denied
2025-07-07 15:09:33.936 [                       datum_protocol_client] ERROR: connect(...) error: Bad file descriptor

BitcoinMechanic commented:
bitcoin@bsd2:~ $ datum_gateway --config=/home/bitcoin/datum/datum_gateway_config.json

 **************************************************************************
 * DATUM Gateway --- Copyright (c) 2024 Bitcoin Ocean, LLC & Jason Hughes *
 * git commit: ac37f191a189c9905861ff2b1c6a9229fd92e1eb                   *
 **************************************************************************

2025-07-07 12:32:09.918 [                         datum_protocol_init]  INFO: Our public keys:
2025-07-07 12:32:09.918 [                   datum_encrypt_log_pubkeys]  INFO: Signing Public Key:     f6e9f8849a402c5c780f1a59ab3c66c6705c6fece1a002e5d88b36295ebbaf35
2025-07-07 12:32:09.918 [                   datum_encrypt_log_pubkeys]  INFO: Encryption Public Key:  e4d23eefac728487d08fc126d0617d53b28dddefe95d0923a574713afeb15f03
2025-07-07 12:32:09.918 [                         datum_protocol_init]  INFO: Pool's public keys: (You should periodically verify that these are what you expect!)
2025-07-07 12:32:09.918 [                   datum_encrypt_log_pubkeys]  INFO: Signing Public Key:     f21f2f0ef0aa1970468f22bad9bb7f4535146f8e4a8f646bebc93da3d89b1406
2025-07-07 12:32:09.918 [                   datum_encrypt_log_pubkeys]  INFO: Encryption Public Key:  f40d032f09a417d94dc068055df654937922d2c89522e3e8f6f0e649de473003
2025-07-07 12:32:09.918 [                                        main]  INFO: Waiting on DATUM server... 15
2025-07-07 12:32:09.918 [                            datum_api_thread]  INFO: API listening on port 7152
2025-07-07 12:32:10.927 [                                        main]  INFO: Waiting on DATUM server... 13
2025-07-07 12:32:11.939 [                                        main]  INFO: Waiting on DATUM server... 12
2025-07-07 12:32:12.949 [                                        main]  INFO: Waiting on DATUM server... 11
2025-07-07 12:32:13.958 [                                        main]  INFO: Waiting on DATUM server... 10
2025-07-07 12:32:14.971 [                                        main]  INFO: Waiting on DATUM server... 9
2025-07-07 12:32:15.977 [                                        main]  INFO: Waiting on DATUM server... 8
2025-07-07 12:32:16.987 [                                        main]  INFO: Waiting on DATUM server... 7
2025-07-07 12:32:17.998 [                                        main]  INFO: Waiting on DATUM server... 6
2025-07-07 12:32:19.008 [                                        main]  INFO: Waiting on DATUM server... 5
2025-07-07 12:32:20.017 [                                        main]  INFO: Waiting on DATUM server... 4
2025-07-07 12:32:21.028 [                                        main]  INFO: Waiting on DATUM server... 3
2025-07-07 12:32:22.037 [                                        main]  INFO: Waiting on DATUM server... 2
2025-07-07 12:32:23.047 [                                        main]  INFO: Waiting on DATUM server... 1
2025-07-07 12:32:24.057 [                                        main]  INFO: Waiting on DATUM server... 0
2025-07-07 12:32:25.068 [                                        main] ERROR: DATUM server connection could not be established.
2025-07-07 12:32:30.146 [               datum_gateway_template_thread] ERROR: Could not fetch new template from localhost:28332!
2025-07-07 12:32:35.277 [                                        main]  WARN: Configured for pooled mining only, and connection lost to DATUM server!  Shutting down Stratum v1 server until DATUM connection reestablished.
2025-07-07 12:32:36.157 [               datum_gateway_template_thread] ERROR: Could not fetch new template from localhost:28332!
2025-07-07 12:32:42.177 [               datum_gateway_template_thread] ERROR: Could not fetch new template from localhost:28332!
2025-07-07 12:32:48.205 [               datum_gateway_template_thread] ERROR: Could not fetch new template from localhost:28332!
PANIC TRIGGERED - 241
2025-07-07 12:32:50.812 [              datum_stratum_v1_socket_server] FATAL: Did not see an initial stratum job after ~25 seconds. Is your node properly setup? Check your config and connectivity and try again!
2025-07-07 12:32:50.812 [                           panic_from_thread] FATAL: ***********************
2025-07-07 12:32:50.812 [                           panic_from_thread] FATAL: *** PANIC TRIGGERED ***
2025-07-07 12:32:50.812 [                           panic_from_thread] FATAL: ***********************
PANIC EXIT.
2025-07-07 12:32:51.107 [                                        main] FATAL: *** PANIC TRIGGERED: EXITING IMMEDIATELY ***

BitcoinMechanic commented:
No, this is what I get:

```
2025-07-07 12:40:56.703 [                         datum_protocol_init]  INFO: Our public keys:
2025-07-07 12:40:56.703 [                   datum_encrypt_log_pubkeys]  INFO: Signing Public Key:     90635b8ce5982a408080eb06b50f1ec88f32ace0c214c0ac247cca68966233cf
2025-07-07 12:40:56.703 [                   datum_encrypt_log_pubkeys]  INFO: Encryption Public Key:  3020dac6ccd9c13ae13e6e1705037695f7b9c8c8a0fdad5ab402ca5de1654a04
2025-07-07 12:40:56.703 [                         datum_protocol_init]  INFO: Pool's public keys: (You should periodically verify that these are what you expect!)
2025-07-07 12:40:56.703 [                   datum_encrypt_log_pubkeys]  INFO: Signing Public Key:     f21f2f0ef0aa1970468f22bad9bb7f4535146f8e4a8f646bebc93da3d89b1406
2025-07-07 12:40:56.703 [                   datum_encrypt_log_pubkeys]  INFO: Encryption Public Key:  f40d032f09a417d94dc068055df654937922d2c89522e3e8f6f0e649de473003
2025-07-07 12:40:56.703 [                                        main]  INFO: Waiting on DATUM server... 15
2025-07-07 12:40:56.704 [                            datum_api_thread]  INFO: API listening on port 7152
2025-07-07 12:40:57.707 [                                        main]  INFO: Waiting on DATUM server... 13
2025-07-07 12:40:58.730 [                                        main]  INFO: Waiting on DATUM server... 12
2025-07-07 12:40:59.737 [                                        main]  INFO: Waiting on DATUM server... 11
2025-07-07 12:41:00.746 [                                        main]  INFO: Waiting on DATUM server... 10
2025-07-07 12:41:01.757 [                                        main]  INFO: Waiting on DATUM server... 9
2025-07-07 12:41:02.767 [                                        main]  INFO: Waiting on DATUM server... 8
2025-07-07 12:41:03.783 [                                        main]  INFO: Waiting on DATUM server... 7
2025-07-07 12:41:04.787 [                                        main]  INFO: Waiting on DATUM server... 6
2025-07-07 12:41:05.797 [                                        main]  INFO: Waiting on DATUM server... 5
2025-07-07 12:41:06.806 [                                        main]  INFO: Waiting on DATUM server... 4
2025-07-07 12:41:07.817 [                                        main]  INFO: Waiting on DATUM server... 3
2025-07-07 12:41:08.826 [                                        main]  INFO: Waiting on DATUM server... 2
2025-07-07 12:41:09.838 [                                        main]  INFO: Waiting on DATUM server... 1
2025-07-07 12:41:10.847 [                                        main]  INFO: Waiting on DATUM server... 0
2025-07-07 12:41:11.858 [                                        main] ERROR: DATUM server connection could not be established.
2025-07-07 12:41:12.265 [               datum_gateway_template_thread]  INFO: NEW NETWORK BLOCK: 0000000000000000000114e217379a884094938054aef4120372f2bd4b1101cb (904496)
2025-07-07 12:41:12.284 [              datum_stratum_v1_socket_server]  INFO: Stratum V1 Server Init complete.
2025-07-07 12:41:12.284 [              datum_stratum_v1_socket_server]  INFO: Empty work send completed. Sent to 0 clients across 0 threads
2025-07-07 12:41:12.284 [               datum_gateway_listener_thread]  INFO: DATUM Socket listener thread active for 'Stratum V1 Server'
2025-07-07 12:41:12.286 [               datum_gateway_template_thread]  INFO: Updating priority stratum job for block 904496: 3.17727914 BTC, 4363 txns, 1505339 bytes (Sent to 0 stratum clients)
2025-07-07 12:41:27.188 [                                        main]  WARN: Configured for pooled mining only, and connection lost to DATUM server!  Shutting down Stratum v1 server until DATUM connection reestablished.
2025-07-07 12:41:27.188 [               datum_stratum_v1_shutdown_all]  INFO: Sent disconnect request for all stratum clients to 0 threads.
```

21M4TW commented:
Ok I now get this in DEBUG mode, which seems to correspond to the issue you also have:
2025-07-07 17:06:10.876 [                   datum_protocol_send_hello] DEBUG: Signing handshake 313 bytes
2025-07-07 17:06:10.876 [                   datum_protocol_send_hello] DEBUG: Sending handshake init (425 bytes)
2025-07-07 17:06:10.881 [                       datum_protocol_client] DEBUG: DATUM Protocol Thread is exiting.
2025-07-07 17:06:11.888 [              datum_protocol_start_connector] DEBUG: Starting DATUM v0.2.3-beta client...
2025-07-07 17:06:11.889 [                                        main]  INFO: Waiting on DATUM server... 13

21M4TW commented:
Ok, so I did some tests, and the issue is in datum_protocol_client. In the first iteration of the main while loop (when i is incremented to 1), epoll_wait times out while connect() is still in progress. By the time the second iteration in the main while loop occurs (when i is incremented to 2), server_out_buf is non-zero, so send() is called, but as connect() is still in progress, connect() returns the error ENOTCONN.

Modifying the epoll_wait line by the following seems to fix the timing issue for me:
`nfds = epoll_wait(epollfd, events, MAX_DATUM_CLIENT_EVENTS, (i>1? 5: 100));  // Wait for 5ms except for first loop`

21M4TW commented:
@BitcoinMechanic  So I created the following branch in my fork with modified code that fixes the issue: https://github.com/21M4TW/datum_gateway/tree/fix_client_connect_in_progress . I tested it and it works for me. Would you like to test it as well before I submit a PR? Thanks

luke-jr commented:
Probably makes sense to just add it to this PR? (Assuming it works)

BitcoinMechanic commented:
I can certainly test - gimme a bit

luke-jr commented:
How about this?

```diff
--- a/src/datum_protocol.c
+++ b/src/datum_protocol.c
@@ -1629,7 +1629,7 @@ void *datum_protocol_client(void *args) {
                                        server_out_buf = 0;
                                }
                        } else {
-                               if (!(errno == EAGAIN || errno == EWOULDBLOCK)) {
+                               if (!(errno == EAGAIN || errno == EWOULDBLOCK || errno == ENOTCONN)) {
                                        pthread_mutex_unlock(&datum_protocol_send_buffer_lock);
                                        break;
                                }
```

BitcoinMechanic commented:
> ```diff
>  || errno == ENOTCONN)) {
> ```

this works
